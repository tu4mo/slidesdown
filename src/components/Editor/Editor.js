import React, { Component, createRef } from 'react'
import PropTypes from 'prop-types'

import {
  StyledWrapper,
  StyledTextarea,
  StyledProgressBar
} from './Editor.style'

import {
  getCurrentLineNumber,
  getSlidesFirstLines,
  getCurrentSlide
} from './utils'

class Editor extends Component {
  static propTypes = {
    isLoading: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    onCursorPositionChange: PropTypes.func,
    onDrop: PropTypes.func,
    progress: PropTypes.number,
    value: PropTypes.string
  }

  editorRef = createRef()

  getCurrentCursorPosition = () => this.editorRef.current.selectionStart

  handleClickAndKeyUp = () => {
    const { onCursorPositionChange, value } = this.props

    const cursorPosition = this.getCurrentCursorPosition()

    const currentLineNumber = getCurrentLineNumber(value, cursorPosition)

    const slides = getSlidesFirstLines(value)
    const slide = getCurrentSlide(slides, currentLineNumber)

    onCursorPositionChange &&
      onCursorPositionChange({
        cursorPosition,
        slide
      })
  }

  handleCursorPositionChange = () => {
    const { onCursorPositionChange } = this.props

    onCursorPositionChange &&
      onCursorPositionChange(this.getCurrentCursorPosition())
  }

  handleDrop = event => {
    const { onDrop } = this.props

    event.preventDefault()
    onDrop && onDrop(event.dataTransfer.files[0])
  }

  render() {
    const { isLoading, onChange, progress, value } = this.props

    return (
      <StyledWrapper>
        <StyledTextarea
          aria-label="Markdown"
          disabled={isLoading}
          onChange={onChange}
          onClick={this.handleClickAndKeyUp}
          onDrop={this.handleDrop}
          onKeyUp={this.handleClickAndKeyUp}
          placeholder="Write markdown here"
          ref={this.editorRef}
          value={value}
        />
        {isLoading && <StyledProgressBar progress={progress} />}
      </StyledWrapper>
    )
  }
}

export default Editor
