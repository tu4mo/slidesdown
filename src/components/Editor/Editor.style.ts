import styled from 'styled-components'

export const StyledWrapper = styled.div`
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  width: 100%;
`

export const StyledTextarea = styled.textarea`
  background-color: var(--color-editor);
  border: 0;
  caret-color: var(--color-purple);
  color: var(--color-light-text);
  font-family: var(--font-monospace);
  font-size: 16px;
  height: 100%;
  left: 0;
  outline: none;
  padding: 1rem;
  position: absolute;
  top: 0;
  resize: none;
  width: 100%;
  -webkit-overflow-scrolling: touch;

  &:disabled {
    opacity: 0.2;
  }

  &::selection {
    background-color: var(--color-purple);
    color: #fff;
  }
`

interface Props {
  progress: number
}

export const StyledProgressBar = styled.div<Props>`
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: center;
  left: 0;
  padding: 2rem;
  position: absolute;
  top: 0;
  width: 100%;

  &::after {
    background-image: linear-gradient(
      90deg,
      var(--color-purple) 0%,
      var(--color-purple) ${(props) => props.progress}%,
      #fff ${(props) => props.progress}%,
      #fff 100%
    );
    border: 2px solid var(--color-purple);
    border-radius: 1rem;
    content: '';
    height: 1rem;
    display: block;
    width: 100%;
  }
`
