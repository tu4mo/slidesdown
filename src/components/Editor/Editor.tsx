import { useRef } from 'react'

import {
  StyledWrapper,
  StyledTextarea,
  StyledProgressBar,
} from './Editor.style'

import {
  getCurrentLineNumber,
  getSlidesFirstLines,
  getCurrentSlide,
} from './utils'

interface Props {
  isLoading?: boolean
  onChange: React.HTMLProps<HTMLTextAreaElement>['onChange']
  onCursorPositionChange?({
    cursorPosition,
    slide,
  }: {
    cursorPosition: number
    slide: number
  }): void
  onDrop?(file: File): void
  progress?: number
  value: string
}

const Editor = ({
  isLoading,
  onChange,
  onCursorPositionChange,
  onDrop,
  progress,
  value,
}: Props) => {
  const editorRef = useRef<HTMLTextAreaElement>(null)

  const getCurrentCursorPosition = (): number =>
    editorRef.current?.selectionStart ?? 0

  const handleClickAndKeyUp = () => {
    const cursorPosition = getCurrentCursorPosition()
    const currentLineNumber = getCurrentLineNumber(value, cursorPosition)
    const slides = getSlidesFirstLines(value)
    const slide = getCurrentSlide(slides, currentLineNumber)

    onCursorPositionChange?.({
      cursorPosition,
      slide,
    })
  }

  const handleDrop = (event: React.DragEvent) => {
    event.preventDefault()
    onDrop?.(event.dataTransfer.files[0])
  }

  return (
    <StyledWrapper>
      <StyledTextarea
        aria-label="Markdown"
        disabled={isLoading}
        onChange={onChange}
        onClick={handleClickAndKeyUp}
        onDrop={handleDrop}
        onKeyUp={handleClickAndKeyUp}
        placeholder="Write markdown here"
        ref={editorRef}
        value={value}
      />
      {isLoading && <StyledProgressBar progress={progress || 0} />}
    </StyledWrapper>
  )
}

export { Editor }
