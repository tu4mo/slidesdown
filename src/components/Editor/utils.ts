interface Slide {
  slide: number
  firstLine: number
}

export const getCurrentLineNumber = (value: string, cursorPosition: number) =>
  value.substr(0, cursorPosition).split('\n').length

export const getSlidesFirstLines = (value = ''): Slide[] =>
  value
    .split('\n')
    .reduce(
      (prev, curr, index) => (curr === '---' ? [...prev, index + 1] : prev),
      [0]
    )
    .map((lines, index) => ({ slide: index, firstLine: lines }))

export const getCurrentSlide = (
  slides: Slide[],
  currentLineNumber: number
): number => {
  const slideObj = [...slides]
    .reverse()
    .find(slide => currentLineNumber > slide.firstLine)

  return (slideObj && slideObj.slide) || 0
}
