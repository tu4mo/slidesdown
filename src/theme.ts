export default {
  breakpoints: {
    md: '576px',
    xl: '768px'
  },
  colors: {
    background: '#edf1f5',
    border: '#e4e4e8',
    dimText: '#a4a4a8',
    lightText: '#646468',
    magenta: '#c86dd7',
    purple: '#3023ae'
  },
  fonts: {
    default: 'Rubik',
    monospace: 'Inconsolata'
  }
}
