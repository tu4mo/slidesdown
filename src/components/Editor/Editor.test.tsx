import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { ThemeProvider } from 'styled-components'

import theme from '../../theme'
import Editor from '../Editor'

it('renders correctly', () => {
  const onChangeMock = jest.fn()

  const tree = renderer
    .create(
      <ThemeProvider theme={theme}>
        <Editor onChange={onChangeMock} value="Test" />
      </ThemeProvider>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('renders correctly with disabled and isLoading', () => {
  const onChangeMock = jest.fn()

  const tree = renderer
    .create(
      <ThemeProvider theme={theme}>
        <Editor isLoading onChange={onChangeMock} progress={50} value="Test" />
      </ThemeProvider>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
