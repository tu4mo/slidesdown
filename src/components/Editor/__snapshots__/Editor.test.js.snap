// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  width: 100%;
}

.c1 {
  border: 0;
  caret-color: #3023ae;
  color: #646468;
  font-family: Inconsolata;
  font-size: 16px;
  height: 100%;
  left: 0;
  outline: none;
  padding: 1rem;
  position: absolute;
  top: 0;
  resize: none;
  width: 100%;
  -webkit-overflow-scrolling: touch;
}

.c1:disabled {
  opacity: 0.2;
}

.c1::selection {
  background-color: #3023ae;
  color: #fff;
}

<div
  className="c0"
>
  <textarea
    className="c1"
    onChange={[MockFunction]}
    onClick={[Function]}
    onDrop={[Function]}
    onKeyUp={[Function]}
    placeholder="Write markdown here"
    value="Test"
  />
</div>
`;

exports[`renders correctly with disabled and isLoading 1`] = `
.c0 {
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  width: 100%;
}

.c1 {
  border: 0;
  caret-color: #3023ae;
  color: #646468;
  font-family: Inconsolata;
  font-size: 16px;
  height: 100%;
  left: 0;
  outline: none;
  padding: 1rem;
  position: absolute;
  top: 0;
  resize: none;
  width: 100%;
  -webkit-overflow-scrolling: touch;
}

.c1:disabled {
  opacity: 0.2;
}

.c1::selection {
  background-color: #3023ae;
  color: #fff;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 0;
  padding: 2rem;
  position: absolute;
  top: 0;
  width: 100%;
}

.c2::after {
  background-image: linear-gradient( 90deg, #3023ae 0%, #3023ae 50%, #fff 50%, #fff 100% );
  border: 2px solid #3023ae;
  border-radius: 1rem;
  content: '';
  height: 1rem;
  display: block;
  width: 100%;
}

<div
  className="c0"
>
  <textarea
    className="c1"
    disabled={true}
    onChange={[MockFunction]}
    onClick={[Function]}
    onDrop={[Function]}
    onKeyUp={[Function]}
    placeholder="Write markdown here"
    value="Test"
  />
  <div
    className="c2"
  />
</div>
`;
